doop.c

```diff
@@ -657,16 +657,68 @@ Magic and tainting are handled.
 =cut
 */
 
+PERL_STATIC_INLINE SV *
+S_do_join_inner(pTHX_ SV *lhs, SV *rhs)
+{
+    PERL_ARGS_ASSERT_DO_JOIN_INNER;
+
+    /* lhs has already had SvGETMAGIC called */
+    if (SvGAMAGIC(rhs)) {
+        SvGETMAGIC(rhs);
+        rhs = sv_mortalcopy_flags(rhs, 0);
+    }
+
+    int applies_flags = 0;
+    /* LD_LIBRARY_PATH=`pwd` perf stat ./perl -e '
+           my @x = ("a" x 100) x 1000;
+           for my $i (0 .. 10000) { my $z = join(",", @x); }'
+     * We speed this up by 20% (about) with a check for both ROK and RMG,
+     * because those must be set if an SV has any overload. */
+    U32 flags = SvFLAGS(lhs)|SvFLAGS(rhs);
+    if (UNLIKELY((flags & (SVf_ROK|SVs_GMG))) &&
+        UNLIKELY(amagic_applies(lhs, concat_amg, applies_flags)
+              || amagic_applies(rhs, concat_amg, applies_flags))) {
+
+        SV *tmpsv = amagic_call(lhs, rhs, concat_amg, 0);
+        assert(tmpsv);
+        if (!tmpsv)
+            croak("panic: do_join_inner expected amagic_call to succeed\n");
+        return tmpsv;
+    }
+
+    /* We speed this up another 10% (about) using sv_catpvn_flags instead of
+     * sv_catsv_nomg(lhs, rhs). Passing in the *delims to reduce calls
+     * to SvPV_nomg_const for the delim made no measurable difference. */
+    STRLEN len;
+    const char * const s = SvPV_nomg_const(rhs, len);
+    sv_catpvn_flags(lhs, s, len, DO_UTF8(rhs) ? SV_CATUTF8 : SV_CATBYTES);
+
+    return lhs;
+}
+
 void
 Perl_do_join(pTHX_ SV *sv, SV *delim, SV **mark, SV **sp)
 {
+    PERL_ARGS_ASSERT_DO_JOIN;
+
     SV ** const oldmark = mark;
+    SV *orig_sv = sv;
+    if (SvGAMAGIC(delim)) {
+        SvGETMAGIC(delim);
+        delim = sv_mortalcopy_flags(delim, 0);
+    }
+    int applies_flags = 0;
+    bool delim_has_concat = amagic_applies(delim, concat_amg, applies_flags);
     I32 items = sp - mark;
     STRLEN len;
     STRLEN delimlen;
-    const char * const delims = SvPV_const(delim, delimlen);
+    const char * const delims = SvPV_nomg_const(delim, delimlen);
 
-    PERL_ARGS_ASSERT_DO_JOIN;
+    /* stringify once and use that unless the delim has concat_amg */
+    if (!delim_has_concat) {
+        delim = newSVpvn_utf8(delims, delimlen, SvUTF8(delim));
```

Comment:
see: https://github.com/Perl/perl5/pull/20503#discussion_r1069270406

Reply:
I'd say that @leonerd's advice was incomplete or perhaps a little overly focused on the point about utf8. `newSVpvn_utf8()` is a macro around `newSVpvn_flags()` that prevents you from passing in the SVs_TEMP flag which would make your code more optimal.  You want to create a new *statement* scoped mortal temporary. The code I showed is the optimal way.  You shouldn't be creating a block scoped temporary variable, you should be creating a statement scoped temporary variable. With your current code this:

```
{
   $str = join (",", $str, $_) for 1..1_000_000;
}
```

will create 1_000_000 temporary values that will not be freed until block exit. With the code I showed, each iteration of join should free up the variables right after the statement completes.

Note
```
#define newSVpvn_utf8(s, len, u) newSVpvn_flags((s), (len), (u) ? SVf_UTF8 : 0)
``` 

If you read the docs for newSVpvn_flags the SVs_TEMP flag results in the low level machinery setting up the returned value as a mortal, which avoids an inefficient call to sv_2mortal() (which you should be using instead of SAVEFREEPV() anyway). You also need to pass in the UTF8 flag to ensure the semantics are correct, so call newSVpvn_flags() directly.

Note the docs for SAVEFREEPV:

```
=item C<SAVEFREESV(SV *sv)>

The refcount of C<sv> will be decremented at the end of
I<pseudo-block>.  This is similar to C<sv_2mortal> in that it is also a
mechanism for doing a delayed C<SvREFCNT_dec>.  However, while C<sv_2mortal>
extends the lifetime of C<sv> until the beginning of the next statement,
C<SAVEFREESV> extends it until the end of the enclosing scope.  These
lifetimes can be wildly different.

Also compare C<SAVEMORTALIZESV>.
```

Note the comparison with sv_2mortal(). SAVEFREESV is one of those things that you should only use if you are doing something with exceptional scoping requirements from the normal case. Like declaring a new variable that needs to survive the full scope of the block. Temporary SV's on the other hand are pretty much universally set up with sv_2mortal() or even better through a newSV.._flags() call with the SVs_TEMP flag.

Also be aware, i typoed originally, and misspelled SVs_TEMP as SVf_TEMP, it is not an SV flag, so it has a different prefix. Sorry about that.

BTW, these stats should give you some insight on why SAVEFREESV is a red flag:

```
$ git grep SAVEFREESV | wc -l
75
$ git grep SVs_TEMP | wc -l
230
$ git grep sv_2mortal | wc -l
765
```

We mortalize about 10 times for every SAVEFREESV.

Reply:
BTW, its maybe worth knowing that a lot of the SV related functions are wrappers passing in specific flags to lower level _flags() based functions. You will find that a lot of our API functions can be decomposed. newSV is the prefix for stuff that creates a new sv. The suffix tells you about the arguments. So another example might sv_cat which has a similar set of suffixes and wrappers to newSV. So newSVpvf results in a newSV from a format string, and sv_catpvf() concatenates a string created by a format.  In a lot, maybe a majority of these cases, the functions are wrappers around a core function with a _flags suffix. It is often useful when optimizing (and learning) to unroll the chain and call directly to the core function with the desired flags and arguments. Once you become familiar with the patterns you can often guess at the names of functions and be correct, they are relatively consistent.